README.md
# Disclaimer

The extension is still under development. You are welcome to test it and submit feedback either via issues or via #postgresql on Libera.chat @nickb. Current state of the extension is the version 0.1.0. Any further changes to the extension will have a corresponding version bump and will come with upgrade script. This means that as of now the extension is safe to install into a database with real data.
# UUID v1 Opclass

This extension provides the `uuid_v1_ops` opclass for indexes over UUID type. This enables a custom sorting order based on the timestamp encoded within the [UUID v1](https://www.ietf.org/rfc/rfc4122.txt).

For an introduction refer to the [wiki page](https://github.com/pgnickb/uuid_v1_ops/wiki)

The benefit of having this over a normal `uuid_ops` is the monotonically increasing nature of the timestamp (under intended workloads). Thus the index in question has values appeneded to its end. This in turn significantly reduces WAL traffic and results in up to 2x performance increase for ingest-heavy workloads.

Otherwise the performance penalty is negligible. 

# Installation

Simply run:
```bash
make
sudo make install
```

As long as you have `pg_config` in your path, this should just work. Drop the `sudo` if your PostgreSQL is running in the userspace.

# Tests

To run the standard regression test suite simply do:

```
make installcheck
```

If you have `pg_prove` available and like `pgTap` then you need to create a database named `test_uuid_v1_ops` and run:
```
make test
```

# Benchmarks

Below are results of a fairly primitive benchmark with a dataset of 40mil UUIDs. This is a "happy" path for our extension, but it's purpose is to demostrate the advantage of using a time UUID in its natural context.

Default behavior:

```SQL
create table u(u uuid);
create index on u(u);

\copy u from ~/temp/40_mil_uuids
```

```
Total WAL generated: 7560 MB 
Time:                120115.153 ms (02:00.115) 
btree FPI %:         100.0
total FPI %:          39.3 
```

With `uuid_v1_ops`:

```SQL
create table v1(u uuid);
create index on v1(u uuid_v1_ops);

\copy v1 from ~/temp/40_mil_uuids
```

```
Total WAL generated: 3818 MB 
Time:                55582.802 ms (00:55.583) 
btree FPI %:         51.23 
total FPI %:          0.01 
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Types of changes

- `Added` for new features. 
- `Changed` for changes in existing functionality. 
- `Deprecated` for soon-to-be removed features. 
- `Removed` for now removed features. 
- `Fixed` for any bug fixes. 
- `Security` in case of vulnerabilities. 

## [Unreleased]

### Added 

- Placeholder


### Fixed

- Placeholder


### Changed

- Placeholder


approve_tests.sh
#!/bin/bash

echo Following contents will be written to the expected:
echo
echo

for x in ./tap_results/sql/*.sql; do
    base=$(basename $x)
    t=./expected/"${base%.sql}.out"
    echo "\unset ECHO" > $t
    echo >> $t
    cat "$x" >> $t
    echo "$t"
    cat "$t"
    echo
done



uuid_v1_ops.h
#ifndef UUID_V1_OPS_H
#define UUID_V1_OPS_H

#include "postgres.h"
#include "nodes/print.h"
#include "utils/builtins.h"
#include "utils/sortsupport.h"
#include "utils/uuid.h"
#include "utils/inet.h"
#include "utils/timestamp.h"

#define UUID_V1_100NS_TO_USEC INT64CONST(10)
#define UUID_V1_TIMESTAMP_LEN 8
#define GREGORIAN_BEGINNING_OFFSET_USEC INT64CONST(13165977600000000)
#define UUID_VERSION_OFFSET 6

#define UUID_V1_GREATEST_SUPPORTED_TIMESTAMP INT64CONST(909171226085477580)
#define UUID_V1_GREATEST_SUPPORTED_INT8 INT64CONST(0xf000000000000000)
#define UUID_V1_LEAST_SUPPORTED_TIMESTAMP -GREGORIAN_BEGINNING_OFFSET_USEC

#define UUID_V1_NODE_OFFSET_A 10
#define UUID_V1_NODE_OFFSET_B 11
#define UUID_V1_NODE_OFFSET_C 12
#define UUID_V1_NODE_OFFSET_D 13
#define UUID_V1_NODE_OFFSET_E 14
#define UUID_V1_NODE_OFFSET_F 15

#define UUID_V1_SEQ_OFFSET 8

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(is_uuid_v1);
Datum		is_uuid_v1(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_cmp);
Datum		uuid_v1_cmp(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_timestamptz);
Datum		uuid_v1_get_timestamptz(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_timestamp_as_int8);
Datum		uuid_v1_get_timestamp_as_int8(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_lt);
Datum		uuid_v1_lt(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_le);
Datum		uuid_v1_le(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_ge);
Datum		uuid_v1_ge(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_gt);
Datum		uuid_v1_gt(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_node_id);
Datum		uuid_v1_get_node_id(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_clock_seq);
Datum		uuid_v1_get_clock_seq(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_variant);
Datum		uuid_v1_get_variant(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_create_from_ts);
Datum		uuid_v1_create_from_ts(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_create_from_int8);
Datum		uuid_v1_create_from_int8(PG_FUNCTION_ARGS);

pg_uuid_t  *uuid_v1_create_from_internal(int64 ts, int16 clock_seq, macaddr *node);
static int64 uuid_v1_get_timestamp_internal(pg_uuid_t *uuid);
static int	uuid_v1_internal_cmp(const pg_uuid_t *arg1, const pg_uuid_t *arg2);


static unsigned int order[UUID_LEN] = {6, 7, 4, 5, 0, 1, 2, 3, 8, 9, 10, 11, 12, 13, 14, 15};

#endif							/* UUID_V1_OPS_H */


uuid_v1_ops.c
#include <uuid_v1_ops.h>

Datum
is_uuid_v1(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);

	PG_RETURN_BOOL(
				   (arg1->data[UUID_VERSION_OFFSET] & 0xf0) == 0x10
		);
}

static int64
uuid_v1_get_timestamp_internal(pg_uuid_t *uuid)
{
	uint64		res;			/* number of 100ns intervals since EPOCH */

	res = ((uint64) ((uuid->data[order[0]]) & 0x0F)) << 56;
	for (int i = 1; i < UUID_V1_TIMESTAMP_LEN; ++i)
	{
		/* this conversion is safe regardless of the endiannes */
		res += ((uint64) uuid->data[order[i]]) << ((UUID_V1_TIMESTAMP_LEN - i - 1) * 8);
	}

	return res;
}

Datum
uuid_v1_get_timestamptz(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	uint64		res;
	uint64		gns;			/* number of 100ns intervals since EPOCH */

	gns = uuid_v1_get_timestamp_internal(arg1);
	res = (gns / UUID_V1_100NS_TO_USEC) - GREGORIAN_BEGINNING_OFFSET_USEC;
	PG_RETURN_TIMESTAMPTZ(res);
}

Datum
uuid_v1_get_timestamp_as_int8(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);

	PG_RETURN_INT64(uuid_v1_get_timestamp_internal(arg1));
}


Datum
uuid_v1_cmp(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_INT32(uuid_v1_internal_cmp(arg1, arg2));
}

static int
uuid_v1_internal_cmp(const pg_uuid_t *arg1, const pg_uuid_t *arg2)
{
	int			res;

	for (int i = 0; i < UUID_LEN; ++i)
	{
		res = arg1->data[order[i]] - arg2->data[order[i]];
		if (res != 0)
			break;
	}

	return res;
}

Datum
uuid_v1_lt(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) < 0);
}

Datum
uuid_v1_le(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) <= 0);
}

Datum
uuid_v1_ge(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) >= 0);
}

Datum
uuid_v1_gt(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) > 0);
}

Datum
uuid_v1_get_node_id(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	macaddr    *result;

	result = (macaddr *) palloc(sizeof(macaddr));

	result->a = arg1->data[UUID_V1_NODE_OFFSET_A];
	result->b = arg1->data[UUID_V1_NODE_OFFSET_B];
	result->c = arg1->data[UUID_V1_NODE_OFFSET_C];
	result->d = arg1->data[UUID_V1_NODE_OFFSET_D];
	result->e = arg1->data[UUID_V1_NODE_OFFSET_E];
	result->f = arg1->data[UUID_V1_NODE_OFFSET_F];

	PG_RETURN_MACADDR_P(result);
}

Datum
uuid_v1_get_clock_seq(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	int16		seq;

	seq = ((int16) ((arg1->data[UUID_V1_SEQ_OFFSET]) & 0x3F)) << 8;
	seq += arg1->data[UUID_V1_SEQ_OFFSET + 1];

	PG_RETURN_INT16(seq);
}

Datum
uuid_v1_get_variant(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	int16		v;

	v = ((int16) ((arg1->data[UUID_V1_SEQ_OFFSET]) & 0xC0)) >> 6;

	PG_RETURN_INT16(v);
}

Datum
uuid_v1_create_from_ts(PG_FUNCTION_ARGS)
{
	TimestampTz ts = PG_GETARG_TIMESTAMPTZ(0);
	int16		clock_seq = PG_GETARG_INT16(1);
	macaddr    *node = PG_GETARG_MACADDR_P(2);

	/*
	 * Convert PostgreSQL epoch usec timestamptz to the UUID v1 Gregorian
	 * epoch 0.1usec This can overflow, so we need to be careful. UUID v1
	 * Timestamp is unsigned and thus can't represent dates before it's epoch.
	 */

	if (unlikely((ts > UUID_V1_GREATEST_SUPPORTED_TIMESTAMP ||
				  ts < UUID_V1_LEAST_SUPPORTED_TIMESTAMP)))
	{
		ereport(ERROR,
				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
				 errmsg("timestamp value out of range for UUID v1")));
	}

	ts = (ts + GREGORIAN_BEGINNING_OFFSET_USEC) * UUID_V1_100NS_TO_USEC;

	PG_RETURN_UUID_P(uuid_v1_create_from_internal(ts, clock_seq, node));
}

Datum
uuid_v1_create_from_int8(PG_FUNCTION_ARGS)
{
	int64		ts = PG_GETARG_INT64(0);
	int16		clock_seq = PG_GETARG_INT16(1);
	macaddr    *node = PG_GETARG_MACADDR_P(2);

	if (unlikely((ts & UUID_V1_GREATEST_SUPPORTED_INT8) != 0))
	{
		ereport(ERROR,
				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
				 errmsg("timestamp value out of range for UUID v1")));
	}


	PG_RETURN_UUID_P(uuid_v1_create_from_internal(ts, clock_seq, node));
}

pg_uuid_t *
uuid_v1_create_from_internal(int64 ts, int16 clock_seq, macaddr *node)
{
	pg_uuid_t  *res;

	res = (pg_uuid_t *) palloc(sizeof(*res));

	/* timestamp low bytes */
	res->data[0] = (uint8) ((ts & 0xff000000) >> 24);
	res->data[1] = (uint8) ((ts & 0xff0000) >> 16);
	res->data[2] = (uint8) ((ts & 0xff00) >> 8);
	res->data[3] = (uint8) ((ts & 0xff));

	/* timestamp mid bytes */
	res->data[4] = (uint8) ((ts & 0xff0000000000) >> 40);
	res->data[5] = (uint8) ((ts & 0xff00000000) >> 32);

	/* timestamp hi bytes and version */
	res->data[6] = (uint8) ((ts & 0x0f00000000000000) >> 56);
	/* append version to the upper MSB of timestamp hi */
	res->data[6] |= (1 << 4);
	res->data[7] = (uint8) ((ts & 0xff000000000000) >> 48);

	/* clock seq hi and reserved */
	res->data[8] = (uint8) ((clock_seq & 0x3f00) >> 8);
	/* append reserver to the clock seq high */
	res->data[8] |= 0x80;
	/* clock seq low */
	res->data[9] = (uint8) (clock_seq & 0xff);

	/* node id */
	res->data[10] = node->a;
	res->data[11] = node->b;
	res->data[12] = node->c;
	res->data[13] = node->d;
	res->data[14] = node->e;
	res->data[15] = node->f;

	return res;
}
